src/nb_foreign.cpp

```diff
@@ -0,0 +1,541 @@
+/*
+    src/nb_foreign.cpp: libnanobind functionality for interfacing with other
+                        binding libraries
+
+    Copyright (c) 2025 Hudson River Trading LLC <opensource@hudson-trading.com>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#if !defined(NB_DISABLE_FOREIGN)
+
+#include "nb_internals.h"
+#include "nb_ft.h"
+#include "nb_abi.h"
+
+NAMESPACE_BEGIN(NB_NAMESPACE)
+NAMESPACE_BEGIN(detail)
+
+// nanobind exception translator that wraps a foreign one
+static void foreign_exception_translator(const std::exception_ptr &p,
+                                         void *payload) {
+    ((pymb_framework *) payload)->translate_exception(&p);
+}
+
+// When learning about a new foreign type, should we automatically use it?
+NB_INLINE bool should_autoimport_foreign(nb_internals *internals_,
+                                         pymb_binding *binding) {
+    return internals_->foreign_import_all &&
+           binding->framework->abi_lang == pymb_abi_lang_cpp &&
+           binding->framework->abi_extra == internals_->foreign_self->abi_extra;
+}
+
+static void nb_type_import_binding(pymb_binding *binding,
+                                   const std::type_info *cpptype) noexcept;
+
+// Callback functions for other frameworks to operate on our objects
+// or tell us about theirs
+
+static void *nb_foreign_from_python(pymb_binding *binding,
+                                    PyObject *pyobj,
+                                    uint8_t convert,
+                                    void (*keep_referenced)(void *ctx,
+                                                            PyObject *obj),
+                                    void *keep_referenced_ctx) noexcept {
+    cleanup_list cleanup{nullptr};
+    auto *td = (type_data *) binding->context;
+    void *result = nullptr;
+    bool ok = nb_type_get(td->type, pyobj,
+                          convert ? uint8_t(cast_flags::convert) : 0,
+                          keep_referenced ? &cleanup : nullptr, &result);
+    if (keep_referenced) {
+        for (uint32_t idx = 1; idx < cleanup.size(); ++idx)
```

Comment:
I am wondering if there are situations where it might be unsafe to deconstruct the temporaries until after the binding call has finished. 🤔  Here, you do this earlier than nanobind would ordinarily do for its own function bindings.